KAFKA-3306: Change metadata response to include required additional fi…

granthenke opened the pull request:
…elds
- Adds boolean type to the protocol
- Allows protocol arrays to be null (optionally)
- Adds support to ask for no topics in the metadata request
- Adds new fields to the Metadata response protocol
- Adds server code to handle new fields
  - Support no-topic metadata requests
  - Track controller id in the metadata cache
  - Check if a topic is considered internal
  - Included rack information if present
  - Include all replicas and ISRs, even if node is down
- Adds test code to test new functionality independent of the client

granthenke committed:
KAFKA-3306: Change metdata response to include required additional fields

- Adds a Boolean type to the protocol
- Allows protocol arrays to be null
- Adds support to ask for no topics in the metadata request
- Adds new fields to the Metadata reponse protocol
- Adds server code to handle new fields
   - Support no-topic metadata requests
   - Track controller id in the metadata cache
   - Check if a topic is considered internal
   - Check if a topic is marked for deletion
   - included rack information if present
- Adds test code to test new functionality independent of the client

granthenke commented:
@junrao @gwenshap @ijuma 
It would be nice to get this into 0.10. It should be the only required changes for existing protocol messages for KIP-4, which would simplify a backport into 0.10.1 when the rest is agreed upon and ready. 

It also adds rack data to the metadata response which goes along with KIP-36 that was added in this release.

ijuma commented:
@granthenke, I agree that it would be great to get this in if there is agreement on the protocol changes. I haven't reviewed it in detail yet, but 2 things:
- The PR needs to be rebased
- I think it would be great to make it possible to ask for an update with `0` topics in cases where we just want to get the updated cluster data. People have run into performance issues when clients have 0 topics, but they ask for _all_ topics because the protocol only supports that.

granthenke commented:
@ijuma I will rebase and look at supporting the idea of `null` (versus empty list) indicating that you don't want to request topic metadata at all. Its a bit nuanced, but its the most compatible and has fewer edge cases. Worst case scenario if someone passes an empty list is performance. This would help solve [KAFKA-3358](https://issues.apache.org/jira/browse/KAFKA-3358).

granthenke commented:
@ijuma I rebased and added the "no topics" functionality.

granthenke commented on clients/src/main/java/org/apache/kafka/common/requests/MetadataRequest.java:
```diff
@@ -25,25 +25,40 @@
+    /**
+     * Constructor to support asking for no topic metadata
+     */
+    public MetadataRequest() {
```
Not sure if this is the best way to do this. I need a different constructor for requesting a `null` list because `null` matches both the list and struct constructors with the same "specificity".

I am open to ideas for a better way to support requesting no topic metadata.

hachikuji replied:
I've been wondering if it would be better to use static factory methods instead of relying on constructors. In that case, you could use something like `MetadataRequest.allTopics()` or something like that.

granthenke replied:
Good suggestion. I like the builder pattern as well. The implementation can be a bit verbose in java, but it can do validation at build time and prevent the telescoping constructor problem.  It may also be able to provide simplified api compatibility.

hachikuji replied:
I also like the builder pattern. It reads nicely and you don't have to care about argument order. On the other hand, it's also easier to forget necessary arguments and it feels kind of silly when the number of arguments is small. One nice thing about factory methods is that you can give them convenient names (e.g. MetadataRequest.allTopicsV1()). Both options are probably better than using the constructors, but I don't think it would be too bad to stick with the current convention for this patch. 

Another option would be to include an explicit flag in the constructor. For example:

``` java
public MetadataRequest(boolean includeAllTopics);
```

We could then have the other constructor which accepts the topic list use the "empty means empty" semantics, and users would have to call this method to get all topics.

granthenke replied:
@hachikuji I do like the idea of a flag to make it explicit. The challenge that poses is compatibility. I think any existing constructor needs to continue to function the way it used to. So empty list would need to continue to signify "all topics", at least for existing constructors.

ijuma replied:
@granthenke request classes are not API so I am not sure why it has to be like that? The official position is that we generate javadoc for classes that are API: https://kafka.apache.org/090/javadoc/index.html (yes, I know it's confusing, I hope to change that so internal classes are all in internal packages).

granthenke replied:
@ijuma I didn't realize compatibility wasn't a concern on these classes. Then we can do whatever we would like here.

granthenke replied:
@hachikuji @ijuma I updated the constructor to take an `allTopics` boolean. I think the changes required to do so may also solve [KAFKA-3358](https://issues.apache.org/jira/browse/KAFKA-3358). It worked nicely with `metadata.needMetadataForAllTopics` in the `NetworkClient`.

gwenshap commented on clients/src/main/java/org/apache/kafka/common/Node.java:
```diff
@@ -102,12 +137,19 @@ public boolean equals(Object obj) {
         if (port != other.port)
             return false;
+        if (rack == null) {
+            if (other.host != null)
```
you mean other.rack?

Also, I'm wondering whether we want to consider a node to be a different node when the rack changed. I guess it depends on how it is used - except it isn't used...

granthenke replied:
Good catch. I just maintained equals meaning all fields match. I suspect thats the safest path to go right now.

gwenshap commented on clients/src/main/java/org/apache/kafka/common/protocol/Protocol.java:
```diff
@@ -87,13 +90,36 @@
 
-    public static final Schema[] METADATA_REQUEST = new Schema[] {METADATA_REQUEST_V0};
-    public static final Schema[] METADATA_RESPONSE = new Schema[] {METADATA_RESPONSE_V0};
+    public static final Schema BROKER_V1 = new Schema(new Field("node_id", INT32, "The broker id."),
```
I noticed you keep using BROKER_V0 anywhere except METADATA, and I think thats the plan going forward as well.
If my understanding is correct, BROKER_V1 is not really a "newer" broker definition (which implies that eventually we'll move everything to use that) but rather a broker definition specific for METADATA request or just a more detailed broker definition. Maybe rename to something less misleading?

gwenshap replied:
And on similar topic (but should be separate JIRA) - the protocol has few different places with BROKER definitions - maybe more consolidation is possible?

granthenke replied:
Yes, my understanding is these "sub-schemas" are supposed to tie directly to single api/protocol to allow them to change independently. I think the fact that a BROKER_V0 was shared was a mistake. I had left the old one being shared just to minimize change. I can break-out and rename a bit to prevent an accidental sharing in the future.

granthenke replied:
I am of the impression that we are not consolidating so that the wire protocols can change independently. However, we want to represent that in Java or Scala object is up to the parser. This decouples protocol from implementation. Though even in implementation we have chosen to duplicate in the past too.

gwenshap commented:
Sorry, noticed a bit late that this PR includes protocol changes that didn't go through a community discussion. Lets make sure there are 3 committers who are ok with the protocol modification?

granthenke committed:
Fix node.equals and separate broker definitions in the protocl

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/MetadataResponse.java:
```diff
@@ -211,20 +236,53 @@ public Cluster cluster() {
         private final Errors error;
         private final String topic;
+        private final boolean isInternal;
+        private final boolean markedForDeletion;
```
Would it make sense to use a bitset for these booleans?

granthenke replied:
Yeah, that could save a byte for each topic. I will mention it during the wire protocol discussion on the next kip call, since this needs to be reviewed/voted.

ijuma replied:
Just noting that this is no longer relevant given the current implementation (for other people reading the PR).

granthenke committed:
Add explicit controll over nullable arrays

granthenke committed:
Merge branch 'trunk' into metadata-changes

# Conflicts:
#	clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java

granthenke committed:
Move controllerId to seperate field

granthenke committed:
Change null behavior of the metadata request

SinghAsDev commented on clients/src/main/java/org/apache/kafka/common/requests/MetadataRequest.java:
```diff
@@ -25,24 +25,42 @@
+    /**
+     * If allTopics is true, then the passed topics will be ignored.
+     */
+    public MetadataRequest(List<String> topics, boolean allTopics) {
```
In current proposal, we are saying a null indicates all topics. Would it be possible to use that here as well?

granthenke replied:
@SinghAsDev The proposal is to use null on the wire protocol, not in the API. In the discussions it was mentioned that being more explicit in the API was favored. 

Beyond that, if you specify `MetadataRequest(null)` java wont know which constructor to use between `MetadataRequest(List<String> topics)` and `MetadataRequest(Struct struct)`.

SinghAsDev replied:
Got it, thanks for the explanation @granthenke.

granthenke replied:
Thanks for the review @SinghAsDev!

granthenke committed:
Remove is_marked_for_deletion flag

granthenke committed:
Merge branch 'trunk' into metadata-changes

# Conflicts:
#	clients/src/main/java/org/apache/kafka/common/protocol/types/ArrayOf.java

granthenke committed:
Cleanup from merge

granthenke committed:
Fix protocol doc

granthenke committed:
Merge branch 'trunk' into metadata-changes

# Conflicts:
#	core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala

granthenke committed:
Change response to include all replicas and isr

granthenke commented:
@junrao @gwenshap @guozhangwang @ewencp @ijuma 
This patch represents the current vote thread for the KIP-4 Metadata changes. Please feel free to review at your convenience.

granthenke committed:
Merge branch 'trunk' into metadata-changes

granthenke committed:
Merge branch 'trunk' into metadata-changes

granthenke commented:
@junrao @gwenshap @guozhangwang @ewencp @ijuma 
Pinging for review. The KIP-4 Metadata changes vote passed and I rebased on trunk.

gwenshap commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -193,15 +193,15 @@ public void updateFetchPositions(Set<TopicPartition> partitions) {
      */
-    public Map<String, List<PartitionInfo>> getTopicMetadata(List<String> topics, long timeout) {
-        if (topics != null && topics.isEmpty())
+    public Map<String, List<PartitionInfo>> getTopicMetadata(List<String> topics, boolean allTopics, long timeout) {
```
Don't we usually add new arguments at the end?

granthenke replied:
Since this was in internals and I needed to update all usages regardless, I located "related" parameters close to each other. I can move to the  end if you prefer.

gwenshap replied:
Its fine. Just checking :)

gwenshap commented on clients/src/main/java/org/apache/kafka/common/Node.java:
```diff
@@ -102,12 +123,17 @@ public boolean equals(Object obj) {
             return false;
+        if (rack == null) {
+            if (other.rack != null)
+                return false;
```
shouldn't we have a case of both null here?

gwenshap commented on clients/src/main/java/org/apache/kafka/common/protocol/types/ArrayOf.java:
```diff
@@ -24,13 +24,29 @@
 
     @Override
     public void write(ByteBuffer buffer, Object o) {
+        if (o == null) {
```
Do we need to validate if this is nullable before writing?

granthenke replied:
I think all of the other types assume validate will be called before calling write and that the written object would be the one returned by the validate call.

Here is a code snippet from `Schema.write(...)`:

``` scala
Object value = f.type().validate(r.get(f));
f.type.write(buffer, value);
```

gwenshap replied:
Got it. Thanks.

gwenshap commented on clients/src/main/java/org/apache/kafka/common/protocol/types/ArrayOf.java:
```diff
@@ -72,6 +94,9 @@ public String toString() {
     public Object[] validate(Object item) {
         try {
+            if (isNullable() && item == null)
+                return null;
```
Do we want to throw an exception with appropriate message if item if null but shouldn't be?

granthenke replied:
It fails further down during the cast the same way it would before nulls were allowed. That exception is caught and re-thrown as a SchemaException. This behavior is validated by `ProtocolSerializationTest.testNulls`.

gwenshap replied:
Yeah, I saw that. I'm suggesting a separate SchemaException with message specific for nulls, to make troubleshooting/debugging a bit easier. Not a big deal though.

granthenke replied:
I can do that. It should probably be done for all Types that don't accept null. I opened [KAFKA-3603](https://issues.apache.org/jira/browse/KAFKA-3603) to track that.

gwenshap replied:
Thanks!

gwenshap replied:
Thanks!

gwenshap replied:
KAFKA-3603 seems to be for something else?

granthenke replied:
Apologies, copy and paste error. Its [KAFKA-3604](https://issues.apache.org/jira/browse/KAFKA-3604).

gwenshap commented on clients/src/main/java/org/apache/kafka/common/protocol/types/Struct.java:
```diff
@@ -307,9 +315,11 @@ public int hashCode() {
-                Object[] arrayObject = (Object []) this.get(f);
-                for (Object arrayItem: arrayObject)
-                    result = prime * result + arrayItem.hashCode();
+                if (this.get(f) != null) {
```
Not sure if it matters, but we will get the exact same hash for empty and null arrays?

granthenke replied:
Good point. We don't have `hashcode` or `equals` defined for `Schema`, `Field` or `Type`...perhaps we should.

gwenshap replied:
Want to create a followup JIRA?

granthenke replied:
Created [KAFKA-3603](https://issues.apache.org/jira/browse/KAFKA-3603) to track that.

ijuma commented on clients/src/main/java/org/apache/kafka/common/protocol/Protocol.java:
```diff
@@ -56,10 +57,13 @@
 
-    public static final Schema BROKER = new Schema(new Field("node_id", INT32, "The broker id."),
+    public static final Schema METADATA_REQUEST_V1 = new Schema(new Field("topics",
+                                                                          new ArrayOf(STRING, true),
```
Personally, I'd introduce a `ArrayOf.nullable` static factory method as I think that would be more readable than the `true` here (since we don't have named arguments in Java).

gwenshap replied:
+1

granthenke replied:
Will do.

gwenshap commented on clients/src/main/java/org/apache/kafka/common/requests/MetadataRequest.java:
```diff
@@ -52,22 +70,33 @@ public AbstractRequestResponse getErrorResponse(int versionId, Throwable e) {
         switch (versionId) {
             case 0:
-                return new MetadataResponse(Collections.<Node>emptyList(), topicMetadatas);
+            case 1:
```
I don't think we can do a fall-through here? Since v1 error response is not a valid v0 response?

granthenke replied:
This can fall through because the constructor supports version 0 and 1. The version is passed as the last parameter. 

This is related to the discussion here: https://github.com/apache/kafka/pull/1095#discussion_r60758124

gwenshap commented on clients/src/main/java/org/apache/kafka/common/requests/MetadataResponse.java:
```diff
@@ -72,13 +76,24 @@
+        super(new Struct(ProtoUtils.responseSchema(ApiKeys.METADATA.id, version)));
 
         this.brokers = brokers;
+        this.controller = getControllerNode(controllerId, brokers);
```
I am bit confused about the use of controller node vs controllerId. 

The protocol requires sending controllerId, which is what we have in most places. But you also find the actual node and provide an API (which isn't used anywhere? not even in tests?) to get the node.

Wondering what was the the plan here.

granthenke replied:
Whenever a broker id is used on the wire protocol, logically this class looks up the broker by id in the brokers list and represents it to the user as a node.  This is very similar logic to how replicas ids are handled in this class. I am not saying its my favorite, but I chose to maintain the existing pattern. 

The controller node is not used yet, but in KIP-4 it may be used to route messages.

ijuma replied:
Generally, I'm against adding unused methods until we actually need to use them. I agree that this may make sense here from a consistency perspective, but even then I would lean towards not having it. If we do have it, then we should use it from a test, at least.

granthenke replied:
This is used in `MetadataRequestTest.testControllerId`. I will look at keeping it around or not. If I do, I will make sure hasController is tested too.

gwenshap replied:
@granthenke has the full scope of KIP-4 though, if he knows it will be used, it is best to add it now (since the whole point was to add the lower-level APIs now)

ijuma replied:
This is an internal class, so methods can be added at any time though. The point was to update the wire protocol now as far as I know.

gwenshap replied:
I don't think it is internal? It isn't in an "internals" package... it is a public class in "common". AFAIK, this makes it public? Obviously it isn't as widely used as KafkaProducer, but my understanding was that anything not in a package called "internals" is public?

ijuma replied:
Today the only public classes are the ones we generate javadoc for (this was confirmed by Jay and Neha). We don't generate the javadoc for the request classes. As you know, I think this is very confusing and I want to change it so that we use  internals packages across the board.

gwenshap replied:
Got it. Thanks for clarifying. If we can modify it at any time, there's really no point in including "may need" features.

granthenke replied:
I would like to leave it for use in unit tests and also with the expectation that it will be used shortly. I will remove any method that is not being used or tested.

gwenshap commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -629,12 +630,15 @@ class KafkaApis(val requestChannel: RequestChannel,
       info("Auto creation of topic %s with %d partitions and replication factor %d is successful"
         .format(topic, numPartitions, replicationFactor))
-      new MetadataResponse.TopicMetadata(Errors.LEADER_NOT_AVAILABLE, topic, java.util.Collections.emptyList())
+      new MetadataResponse.TopicMetadata(Errors.LEADER_NOT_AVAILABLE, topic, common.Topic.isInternal(topic),
```
I'm wondering if we can push down the work of figuring out if a topic is internal to TopicMetadata constructor. It will clean up the API and TopicMetadata has all the information it needs to find out.

SinghAsDev replied:
+1

granthenke replied:
Thats an interesting idea. Let me take a look and report back.

ijuma replied:
Not sure about that. The request classes shouldn't really have any logic IMO. They should just be a way to serialize/deserialize to the Kafka protocol. We could have a util method somewhere that does this. Or at least a static factory method (instead of constructor) in the request classes as a pragmatic compromise.

granthenke replied:
This is related to this discussion too: https://github.com/apache/kafka/pull/1095#discussion_r60764126

Right now some these classes have quite a bit of logic and don't expose the "raw" information. I agree that it would be nice to have some layering here where the logic is handled/exposed elsewhere. I don't want to impose to much change on existing logic in this patch though either.

ijuma replied:
This is a bit different. There are 3 levels, in a sense:
1. Wire protocol
2. Wrapping the wire protocol into domain model classes
3. Having business logic like what topics are internal in the request classes (or calling helper methods for that from within the request classes)

As we discussed previous @granthenke, I think we should really do `1`, but we do `2` in a number of places. This suggestion is `3`. In any case, I'm OK if we do it in a static factory method as a starting point (and we can move it elsewhere in the future in a separate PR maybe).

gwenshap commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -702,22 +715,30 @@ class KafkaApis(val requestChannel: RequestChannel,
-      Seq.empty[MetadataResponse.TopicMetadata]
-    else
-      getTopicMetadata(authorizedTopics, request.securityProtocol)
+    val errorUnavailableEndpoints = requestVersion == 0
```
Mind adding a comment here, something along the lines of:
"In version 0, we returned an error when brokers with replicas were unavailable, while in higher versions we simply didn't include the broker in the list we returned"
(or is it vice-versa? anyway, its non-obvious and need a comment IMO)

granthenke replied:
Will add the comment.

SinghAsDev commented on clients/src/main/java/org/apache/kafka/common/protocol/Protocol.java:
```diff
@@ -56,10 +57,13 @@
+                                                                          new ArrayOf(STRING, true),
+                                                                          "An array of topics to fetch metadata for. If the topics array is null fetch metadata for all topics."));
+
+    public static final Schema METADATA_BROKER_V0 = new Schema(new Field("node_id", INT32, "The broker id."),
```
NIT: I think BROKER or BROKER_INFO or BROKER_METADATA would be better.

granthenke replied:
The idea is that this is the broker used with the metadata response. All objects tied to the metadata response are prefixed with METADATA. There are other broker objects in this file that I would not want to be confused.

SinghAsDev replied:
Ok, thanks for the explanation.

gwenshap commented on core/src/test/scala/unit/kafka/server/HighwatermarkPersistenceTest.scala:
```diff
@@ -47,9 +50,11 @@ class HighwatermarkPersistenceTest {
-    val zkUtils = EasyMock.createMock(classOf[ZkUtils])
-    EasyMock.replay(zkUtils)
-    
+    val zkClient = EasyMock.createNiceMock(classOf[ZkClient])
```
Not 100% sure about changes to this test. Random cleanup or related to KIP-4?

granthenke replied:
This is left over from changes required when leveraging zookeeper to track deletes. I left it because it looks like its closer to what was actually trying to be mocked based on the comment. I can keep it or leave it.

SinghAsDev commented on clients/src/main/java/org/apache/kafka/common/protocol/Protocol.java:
```diff
@@ -87,13 +91,34 @@
+                                                      new Field("host", STRING, "The hostname of the broker."),
+                                                      new Field("port", INT32,
+                                                        "The port on which the broker accepts requests."),
+                                                      new Field("rack", NULLABLE_STRING, "The rack of the broker."));
```
Should we mention effect of a null string as rack here?

granthenke replied:
What message would you suggest?

SinghAsDev replied:
Something that would capture the following, need not be this verbose though.

> If one or more brokers does NOT have rack information
> For auto topic creation, AdminUtils.assignReplicasToBrokers will create the same assignment as the current implementation (as if no broker has the rack information) and continue with topic creation. This allows auto topic creation to work when doing rolling upgrade.
> For command line tools (TopicCommand and ReassignPartitionsCommand), an exception will be thrown. This will alert the user that a broker may be misconfigured. An additional command line argument --ignore-racks can be supplied to suppress such error and continue with topic creation ignoring all rack information.

granthenke replied:
That information is useful for rack configuration on the broker. I am not sure if it needs to be in the protocol documentation for the metadata response.

gwenshap commented on core/src/test/scala/unit/kafka/server/MetadataCacheTest.scala:
```diff
@@ -197,7 +197,7 @@ class MetadataCacheTest {
     cache.updateCache(15, updateMetadataRequest)
 
-    val topicMetadatas = cache.getTopicMetadata(Set(topic), SecurityProtocol.PLAINTEXT)
+    val topicMetadatas = cache.getTopicMetadata(Set(topic), SecurityProtocol.PLAINTEXT, errorUnavailableEndpoints = true)
```
Aren't we adding a test for the results with errorUnavailableEndpoints = false?

gwenshap replied:
ok, I saw we are testing both request versions below. I still find it a bit weird it is missing from the cache tests, since it is added functionality for the cache - but up to you.

granthenke replied:
I can add those tests here too.

gwenshap commented on core/src/test/scala/unit/kafka/server/MetadataRequestTest.scala:
```diff
@@ -0,0 +1,156 @@
+  }
+
+  @Test
+  def testControllerId() {
```
Love it!

gwenshap commented on core/src/test/scala/unit/kafka/server/MetadataRequestTest.scala:
```diff
@@ -0,0 +1,156 @@
+    TestUtils.createTopic(zkUtils, "t1", 3, 2, servers)
+    TestUtils.createTopic(zkUtils, "t2", 3, 2, servers)
+
+    val metadataResponse = sendMetadataRequest(noMetadataRequest, 1)
```
Do you want to also validate that V0 request with null still returns all topics?

gwenshap replied:
I also don't see us checking for V1 all-topics explicitly. Do you want to add something?

granthenke replied:
Sure I will add some more test cases.

gwenshap commented on core/src/test/scala/unit/kafka/server/SimpleFetchTest.scala:
```diff
@@ -69,8 +69,10 @@ class SimpleFetchTest {
     // create nice mock since we don't particularly care about zkclient calls
-    val zkUtils = EasyMock.createNiceMock(classOf[ZkUtils])
-    EasyMock.replay(zkUtils)
+    val zkClient = EasyMock.createNiceMock(classOf[ZkClient])
```
Not sure how this is related.

granthenke replied:
This is left over from changes required when leveraging zookeeper to track deletes. I left it because it looks like its closer to what was actually trying to be mocked based on the comment. I can keep it or leave it.

ijuma replied:
I agree that these changes (and also in the other file) are weird to include here.

SinghAsDev commented on clients/src/main/java/org/apache/kafka/common/protocol/types/Struct.java:
```diff
@@ -148,6 +148,14 @@ public String getString(String name) {
         return (String) get(name);
     }
 
+    public Boolean getBoolean(Field field) {
```
Just thinking it loud here, it seems we follow getter convention here, we really do not have a common convention across the codebase. However, it is not too bad, as this is not end-user. I think user facing classes/interfaces do not follow getter convention though.

granthenke replied:
This class uses get for all other methods so I followed that.

ijuma replied:
`boolean` is a keyword, that's why.

granthenke replied:
That too.

SinghAsDev replied:
Yea, I was just curious about reasoning, not specifically for the method.

SinghAsDev commented on clients/src/main/java/org/apache/kafka/common/requests/MetadataRequest.java:
```diff
@@ -52,22 +70,33 @@ public AbstractRequestResponse getErrorResponse(int versionId, Throwable e) {
     }
 
+    public boolean hasTopics() {
+        return topics != null;
```
shouldn't we check something like `topics != null && !topics.isEmpty()` or maybe we just want to rename the method to indicate we are indeed checking for topics to be non-null.

SinghAsDev commented on clients/src/main/java/org/apache/kafka/common/requests/MetadataResponse.java:
```diff
@@ -32,14 +31,18 @@
     private static final String PORT_KEY_NAME = "port";
+    private static final String RACK_KEY_NAME = "rack";
+
+    private static final String CONTROLLER_ID_KEY_NAME = "controllerId";
```
Would it make sense to be consistent with other key names? Have something like "controller_id".

SinghAsDev commented on core/src/test/scala/unit/kafka/utils/TestUtils.scala:
```diff
@@ -953,7 +953,7 @@ object TestUtils extends Logging {
 
     val values = (0 until numMessages).map(x => s"test-$x")
-    
+
```
NIT: Maybe we can leave this file out if there are no other changes?

ijuma commented on clients/src/main/java/org/apache/kafka/common/protocol/types/Type.java:
```diff
@@ -59,6 +59,47 @@ public boolean isNullable() {
+
+        @Override
+        public Object read(ByteBuffer buffer) {
+            Byte value = buffer.get();
```
This should be `byte` instead of `Byte`.

granthenke replied:
Will fix.

ijuma commented on clients/src/main/java/org/apache/kafka/common/protocol/types/Type.java:
```diff
@@ -59,6 +59,47 @@ public boolean isNullable() {
+     * the value of 0 to represent false, and 1 to represent true.
+     *
+     * If for some reason a value that is not 0 or 1 is read,
+     * then any non-zero value will return true.
```
What is the reasoning for accepting any non-zero value as `true`? It seems more error-prone IMO. If any value outside of `0` or `1` are used, it's probably a mistake.

granthenke replied:
I implemented it this way to match the existing behavior of StopReplica requests delete_partitions boolean. That way this type could be used there as well without a version bump, and all the "booleans" would behave the same. I planned to update StopReplicaRequest after this patch based on the decisions here. 

Here is the relevant code: https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/requests/StopReplicaRequest.java#L83

ijuma replied:
Legacy code is annoying. ;)

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/MetadataRequest.java:
```diff
@@ -25,24 +25,42 @@
+    }
+
+    /**
+     * If allTopics is true, then the passed topics will be ignored.
```
Do we want to ignore or it would it be better to validate it (ie `topics` must be empty or null for that case).

granthenke replied:
Its a bit of a gray area. We introduced the boolean to be explicit and be the single deciding factor for all topics. If we have this validation, its actually the topics list that really matters again and we essentially have 2 toggles to enable all topics.

When validating would we require null & true to get all topics and fail in all other cases in order to match closely to the wire protocols behavior?

ijuma replied:
The idea is to avoid mistakes. So, if someone gets the boolean logic inverted and passes some topics, then the validation would find that.

ijuma replied:
By the way, I agree that this is a bit subjective.

granthenke replied:
I will play around with some options and run it by you.

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/MetadataResponse.java:
```diff
@@ -149,23 +178,44 @@ public MetadataResponse(Struct struct) {
+        for (Node broker : brokers) {
+            if (broker.id() == controllerId) {
+                controller = broker;
+                break;
```
You can just return the controller here and then you don't need the `controller` variable or break (in Scala, I don't like to use `return`, but if one is using `break` already, then it's at a similar level IMO).

granthenke replied:
I am okay with that.

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/MetadataResponse.java:
```diff
@@ -72,13 +76,24 @@
+    /**
+     * Constructor for the latest version
+     */
+    public MetadataResponse(List<Node> brokers, int controllerId, List<TopicMetadata> topicMetadata) {
```
If possible, Jun generally suggests that we have one constructor per version with older version constructors deprecated. Seems like we could do that here right? The older version would not have `controllerId`.

granthenke replied:
@ijuma I had maintained the old constructor for compatibility in an older version of the patch. But in previous reviews I was told I did not need to maintain it: https://github.com/apache/kafka/pull/1095#discussion_r58058004

ijuma replied:
Sorry for the confusion. If we don't need to use the older version, we can remove it. However, if we still need to use it, then it's better to keep a constructor for the older version instead of having a version parameter. That's based on Jun's previous reviews, I actually prefer the current approach as we don't need to call deprecated constructors. Anyway, if Gwen is fine with this, maybe we can leave as is.

granthenke replied:
These constructors are generally only used in the broker. I like passing in the version because then one constructor can handle all versions and the broker is guaranteed to send the correct version back without a bunch of if/switch logic.

There could be a case in the future where a separate constructor is required to maintain compatibility. But if I can avoid many constructors I would prefer it.

ijuma replied:
I know what you mean, I did the same in a previous PR, but Jun made me change it. ;) As I said, let's see what Gwen thinks.

gwenshap replied:
I actually like the deprecated constructors because we get explicit compile warnings when we bump versions and don't accidentally forget to update some of the places where they are called. 

If you are not convinced that having warnings to help with bumps is useful, I won't make you change it ;)

granthenke replied:
@gwenshap I think it actually ends up being a bit of "crying wolf" since the brokers would be required to use the old constructors to support the old versions. We would throw, and likely suppress, the deprecation warnings in that case. I think unit tests are the best way to verify this behavior. I will work on increasing the version coverage on this patch and consider updating other apis at a later time.

gwenshap replied:
Sounds good!

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/MetadataResponse.java:
```diff
@@ -87,15 +102,22 @@ public MetadataResponse(List<Node> brokers, List<TopicMetadata> topicMetadata) {
         struct.set(BROKERS_KEY_NAME, brokerArray.toArray());
 
+        if (struct.hasField(CONTROLLER_ID_KEY_NAME))
+            struct.set(CONTROLLER_ID_KEY_NAME, controllerId);
```
I think it would be useful to add a comment here and when we set other new fields (`RACK_KEY_NAME`, `IS_INTERNAL_KEY_NAME`, etc.) saying stating the version they were added in (we do that in other request/response classes).

granthenke replied:
I have added comments that they only exist in v1+

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/MetadataResponse.java:
```diff
@@ -130,15 +152,22 @@ public MetadataResponse(Struct struct) {
             Struct topicInfo = (Struct) topicInfos[i];
             Errors topicError = Errors.forCode(topicInfo.getShort(TOPIC_ERROR_CODE_KEY_NAME));
             String topic = topicInfo.getString(TOPIC_KEY_NAME);
+            boolean isInternal = topicInfo.hasField(IS_INTERNAL_KEY_NAME) ? topicInfo.getBoolean(IS_INTERNAL_KEY_NAME) : false;
```
Hmm, if we don't have that field, then we don't know if it's internal or not, right?

granthenke replied:
This is challenging because we use the same class for both V0 and V1 responses. In the common case if you have the new version of the class you would be using V1 protocol, but thats not always true. In the case of new fields I tried to pick a safe default. Since V0 is not aware of internal topics I defaulted to false. Do you have thoughts on a better approach?

ijuma replied:
OK, maybe add a comment in this case as it's not obvious.

granthenke replied:
I can do that. I will include explanation in the version comments suggested earlier.

ijuma commented on clients/src/test/java/org/apache/kafka/common/protocol/types/ProtocolSerializationTest.java:
```diff
@@ -42,21 +43,27 @@ public void setup() {
     public void testSimple() {
+        check(Type.BOOLEAN, false);
+        check(Type.BOOLEAN, true);
+        check(Type.BOOLEAN, false);
```
Do we need to check `false` twice?

granthenke replied:
No. Will fix.

ijuma commented on clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java:
```diff
@@ -278,22 +282,22 @@ private AbstractRequestResponse createListOffsetResponse() {
+        allTopicMetadata.add(new MetadataResponse.TopicMetadata(Errors.NONE, "topic1", false,
                 Arrays.asList(new MetadataResponse.PartitionMetadata(Errors.NONE, 1, node, replicas, isr))));
-        allTopicMetadata.add(new MetadataResponse.TopicMetadata(Errors.LEADER_NOT_AVAILABLE, "topic2",
+        allTopicMetadata.add(new MetadataResponse.TopicMetadata(Errors.LEADER_NOT_AVAILABLE, "topic2", false,
```
Shall we add a topic where the boolean is `true`?

granthenke replied:
Sure we can.

ijuma commented on core/src/main/scala/kafka/admin/AdminClient.scala:
```diff
@@ -83,7 +83,7 @@ class AdminClient(val time: Time,
 
   private def findAllBrokers(): List[Node] = {
-    val request = new MetadataRequest(List[String]())
+    val request = new MetadataRequest(List[String](), true)
```
Should we just pass `null` instead of empty list (like in `RequestResponseTest`)?

granthenke replied:
I think It depends on the choice in earlier comments here: https://github.com/apache/kafka/pull/1095#discussion_r60764750

ijuma commented on core/src/main/scala/kafka/common/Topic.scala:
```diff
@@ -62,4 +62,8 @@ object Topic {
 
+  def isInternal(topic: String): Boolean = {
+    INTERNAL_TOPICS.contains(topic)
+  }
```
Nitpick: no need for braces.

granthenke replied:
Sure. Will remove.

ijuma commented on clients/src/main/java/org/apache/kafka/common/Node.java:
```diff
@@ -102,12 +123,17 @@ public boolean equals(Object obj) {
     @Override
     public String toString() {
-        return host + ":" + port + " (id: " + idString + ")";
+        return host + ":" + port + " (id: " + idString + " rack: " + rack + ")";
```
We are using this `toString` in a number of log statement, not sure the rack information is that relevant (the fact that `toString` is used for both debug and more user-facing string is a pain). @hachikuji, you cleaned up the logs, what do you think?

granthenke committed:
Adress smaller review feedback

ijuma commented on core/src/main/scala/kafka/server/MetadataCache.scala:
```diff
@@ -40,16 +40,22 @@ import org.apache.kafka.common.requests.{MetadataResponse, UpdateMetadataRequest
 
-  private def getAliveEndpoints(brokers: Iterable[Int], protocol: SecurityProtocol): Seq[Node] = {
+  // filterUnavailableEndpoints exists to support v0 MetadataResponses
+  private def getEndpoints(brokers: Iterable[Int], protocol: SecurityProtocol, filterUnavailableEndpoints: Boolean): Seq[Node] = {
```
This method is the main hotspot when it comes to the performance of metadata requests, so we need to be careful about adding additional logic here. I should have added a comment saying that, maybe you could do that.

Under the assumption that unavailable nodes are rare, it seems like this change is safe.

granthenke replied:
I will add the comment.

ijuma commented on core/src/main/scala/kafka/server/MetadataCache.scala:
```diff
@@ -151,8 +158,17 @@ private[server] class MetadataCache(brokerId: Int) extends Logging {
 
+  def getControllerId: Int = {
+    controllerId.getOrElse(MetadataResponse.NO_CONTROLLER_ID)
+  }
```
Why don't we return `Option[Int]` here?

granthenke replied:
Yeah I should return the option here and use `MetadataResponse.NO_CONTROLLER_ID` KafkaApis.

ijuma commented on core/src/main/scala/kafka/server/MetadataCache.scala:
```diff
@@ -151,8 +158,17 @@ private[server] class MetadataCache(brokerId: Int) extends Logging {
     inWriteLock(partitionMetadataLock) {
+      controllerId =
+        updateMetadataRequest.controllerId match {
+          case id if id < 0 => None
```
Is there a reason why we don't check that the value is the same as `NO_CONTROLLER_ID`?

granthenke replied:
Since any negative value is invalid I map it to None.

ijuma replied:
The issue with this approach is that we accept instead of failing when we receive invalid values. In some cases, this can lead to bugs being missed.

ijuma commented on core/src/main/scala/kafka/server/MetadataCache.scala:
```diff
@@ -151,8 +158,17 @@ private[server] class MetadataCache(brokerId: Int) extends Logging {
   def updateCache(correlationId: Int, updateMetadataRequest: UpdateMetadataRequest) {
     inWriteLock(partitionMetadataLock) {
+      controllerId =
+        updateMetadataRequest.controllerId match {
```
Formatting nit: this could be on the previous line.

ijuma commented on core/src/test/scala/unit/kafka/server/MetadataRequestTest.scala:
```diff
@@ -0,0 +1,156 @@
+
+  @Before
+  override def setUp() {
+    super.setUp()
```
Seems unnecessary since the superclass already defines it?

ijuma commented on core/src/test/scala/unit/kafka/server/MetadataRequestTest.scala:
```diff
@@ -0,0 +1,156 @@
+    val controllerId = controllerServer.apis.brokerId
+
+    assertEquals("Controller id should match the active controller",
+      controllerServer.apis.brokerId, controllerId)
```
This is doing the same thing as the `controllerId` line. I think you meant to get the controller id from `metadataResponse`.

ijuma commented on core/src/test/scala/unit/kafka/server/MetadataRequestTest.scala:
```diff
@@ -0,0 +1,156 @@
+
+    assertNotEquals("Controller id should switch to a new broker", controllerId, controllerId2)
+    assertEquals("Controller id should match the active controller after failover",
+      controllerServer.apis.brokerId, metadataResponse.controller.id)
```
Don't you mean to use `controllerServer2` and `metadataResponse2`?

ijuma commented on core/src/test/scala/unit/kafka/server/MetadataRequestTest.scala:
```diff
@@ -0,0 +1,156 @@
+  @Test
+  def testRack() {
+    val metadataResponse = sendMetadataRequest(allMetadataRequest, 1)
+    // Validate rack matches whats set in generateConfigs() above
```
Typo, `what's`.

ijuma commented on core/src/test/scala/unit/kafka/server/MetadataRequestTest.scala:
```diff
@@ -0,0 +1,156 @@
+    // Validate rack matches whats set in generateConfigs() above
+    metadataResponse.brokers().asScala.foreach { broker =>
+      assertEquals("Rack information should match config", s"rack/${broker.id}", broker.rack())
+    }
```
Nitpicks: `()` not needed in `brokers` and `rack`.

ijuma replied:
There's a few other examples in this file.

ijuma commented on core/src/test/scala/unit/kafka/server/MetadataRequestTest.scala:
```diff
@@ -0,0 +1,156 @@
+
+    TestUtils.waitUntilTrue({ () =>
+      val response = sendMetadataRequest(new MetadataRequest(List(replicaDownTopic).asJava), 1)
+      val metadata = response.topicMetadata().asScala.head.partitionMetadata().asScala.head
```
Can we rely on the ordering of the metadata in the response and of the partition metadata?

ijuma replied:
Same question for a couple of other cases like this below.

granthenke replied:
Since I am asking for only 1 topic and that topic has only 1 partition it shouldn't matter here.

ijuma commented on core/src/test/scala/unit/kafka/server/MetadataRequestTest.scala:
```diff
@@ -0,0 +1,156 @@
+    }.get
+    downNode.shutdown()
+
+    TestUtils.waitUntilTrue({ () =>
```
Nitpick: for cases like this, I think it's more readable to have the opening brace after `=>`. For cases where the opening brace can replace the opening parenthesis, it makes sense to position the brace like this case. A bit subjective though.

ijuma commented on core/src/test/scala/unit/kafka/server/BaseRequestTest.scala:
```diff
@@ -0,0 +1,96 @@
+import org.junit.Before
+
+abstract class BaseRequestTest extends KafkaServerTestHarness {
+  val numBrokers = 3
```
How do we know that there are always 3 brokers? I guess the intent is that subclasses will use this in `generateConfigs` but that seems easy to miss. We could define this as a def and let the subclasses define it. Or we could implement `generateConfigs` here and call an abstract method `generateConfig(brokerId)` or something.

granthenke replied:
I moved the configuration to the BaseRequestTest and provided a way to override the properties in the subclass.

ijuma commented on core/src/test/scala/unit/kafka/server/BaseRequestTest.scala:
```diff
@@ -0,0 +1,96 @@
+    } finally {
+      plainSocket.close()
+    }
+  }
```
The 3 methods above seem to be used in many tests. Would it make sense to have a utility class or trait that people can mix-in?

granthenke replied:
There is definitely a better way to share this test code across other tests. I would like to do this cleanup, but would you mind if I open a separate jira to track that and do it shortly after this patch?

ijuma replied:
Sure.

ijuma commented on core/src/test/scala/unit/kafka/server/BaseRequestTest.scala:
```diff
@@ -0,0 +1,96 @@
+    * Serializes and send the request to the given api. A ByteBuffer containing the response is returned.
+    */
+  def send(request: AbstractRequest, apiKey: ApiKeys, version: Short): ByteBuffer = {
+    correlationId = correlationId + 1
```
Nitpick: `correlationId += 1`

ijuma commented on core/src/test/scala/unit/kafka/server/BaseRequestTest.scala:
```diff
@@ -0,0 +1,96 @@
+    val response = requestAndReceive(serializedBytes)
+
+    val responseBuffer = ByteBuffer.wrap(response)
+    val responseHeader = ResponseHeader.parse(responseBuffer)
```
This seems to be unused?

granthenke replied:
It is unused, I just need to call parse to ensure it parses correctly and moves the buffer forward. I can remove the val assignment though.

ijuma replied:
Sounds good to me. Also, please add a comment as side-effects like that are not obvious.

ijuma commented:
@granthenke thanks for the PR. I left a few minor comments, looks good otherwise. I think one thing worth paying special attention to is the `errorUnavailableEndpoints` change, but it seems like Gwen has already done that.

granthenke committed:
Merge branch 'trunk' into metadata-changes

granthenke committed:
Use nullable static factory method in ArrayOf

granthenke committed:
Use static method for all topics request

granthenke committed:
Add MetadataCacheTests when errorUnavailableEndpoints is false

granthenke committed:
Add all topics request test

granthenke commented:
@gwenshap  @ijuma @SinghAsDev @hachikuji 
I updated the patch based on the reviews. I added more tests, and some static factory methods. 

Most notably I changed from using and allTopics boolean to using null and a static method. It required a little bit of re-work on internal apis but looks reasonable to me. If you think the allTopics boolean is better don't hesitate to weigh in. The specific commit with that change is here: https://github.com/apache/kafka/pull/1095/commits/52aa74d479e39479a9c79d09b0e0fe77391fa219

granthenke committed:
Merge branch 'trunk' into metadata-changes

ijuma commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -183,25 +183,26 @@ public void updateFetchPositions(Set<TopicPartition> partitions) {
-        if (topics != null && topics.isEmpty())
+    public Map<String, List<PartitionInfo>> getTopicMetadata(MetadataRequest request, long timeout) {
+        // Save the round trip if no topics are requested. In the case of null, all topics are being requested.
+        if (request.topics() != null && request.topics().isEmpty())
```
We can call `isAllTopics` on `request` and remove the comment about `null`

ijuma commented on core/src/test/scala/unit/kafka/server/BaseRequestTest.scala:
```diff
@@ -0,0 +1,105 @@
+  val numBrokers = 3
+  private var correlationId = 0
+
+  def propertyOverrides(properties: Properties): Properties = properties
```
I think it's important to be clear about the expected usage of methods like this. Either we return a new instance or we mutate the passed instance. Given how `Properties` are generally used, isn't it better to just not return `Properties`?

ijuma commented on core/src/test/scala/unit/kafka/server/BaseRequestTest.scala:
```diff
@@ -0,0 +1,105 @@
+    servers.find { server =>
+      val state = server.brokerState.currentState
+      state != NotRunning.state && state != BrokerShuttingDown.state
+    }.get.socketServer
```
Shall we add a helpful message via `getOrElse`?

ijuma commented on core/src/test/scala/unit/kafka/server/MetadataRequestTest.scala:
```diff
@@ -0,0 +1,165 @@
+    val metadataResponse = sendMetadataRequest(MetadataRequest.allTopics(), 1)
+
+    assertEquals("Controller id should match the active controller",
+      controllerServer.apis.brokerId, metadataResponse.controller.id)
```
the second argument to `assertEquals` could be `controllerId` right?

ijuma commented on core/src/test/scala/unit/kafka/server/MetadataRequestTest.scala:
```diff
@@ -0,0 +1,165 @@
+  @Test
+  def testControllerId() {
+    val controllerServer = servers.find(_.kafkaController.isActive()).get
+    val controllerId = controllerServer.apis.brokerId
```
This is kind of weird, we should probably get the broker id from the config (ie `controllerServer.config.brokerId`) or add a method to `KafkaServer` that returns the broker id.

granthenke replied:
Will change to config

ijuma commented on core/src/test/scala/unit/kafka/server/MetadataRequestTest.scala:
```diff
@@ -0,0 +1,165 @@
+    TestUtils.waitUntilTrue(() => {
+      val metadataResponse2 = sendMetadataRequest(MetadataRequest.allTopics(), 1)
+      controllerServer2.apis.brokerId == metadataResponse2.controller.id
+    }, "Controller id should match the active controller after failover", 5000)
```
It's a bit weird that we assert that failover happened and then we wait until failover happens. Shouldn't the `waitUntilTrue` be before anything else?

granthenke replied:
This is because the failover is basically immediate when looking up the servers directly via:

``` scala
servers.find(_.kafkaController.isActive()).get
```

But propagating that state to the MetadataCache and therefore the MetadataResponse could take a bit of time and requires the `TestUtils.waitUntilTrue`

ijuma replied:
Thanks for the explanation.

ijuma commented on core/src/test/scala/unit/kafka/server/MetadataRequestTest.scala:
```diff
@@ -0,0 +1,165 @@
+    assertTrue("Response should have one topic", v1MetadataResponse.topicMetadata.size == 1)
+    val v1PartitionMetadata = v1MetadataResponse.topicMetadata.asScala.head.partitionMetadata.asScala.head
+    assertTrue("PartitionMetadata should have no errors", v1PartitionMetadata.error == Errors.NONE)
+    assertTrue(s"Response should have $replicaCount replicas", v1PartitionMetadata.replicas.size == replicaCount)
```
We should also include an assert for the field we can use to know that a replica is down.

granthenke replied:
Will add a check to confirm the downed broker is not in the brokers list

ijuma commented:
Thanks @granthenke. Aside from a few minor comments, LGTM. I started a system tests build here:

https://jenkins.confluent.io/job/system-test-kafka-branch-builder/415/console

granthenke committed:
Use isAllTopics & Improve tests

granthenke commented:
@ijuma I updated the patch based on your feedback

ijuma commented:
Thanks @granthenke, LGTM. @gwenshap, do you want to take a look as well?

gwenshap commented:
Looks perfect. I was waiting for you to finish nitpicking tests ;)
